src/Http/Routing/src/MapAction/MapActionEndpointRouteBuilderExtensions.cs

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.AspNetCore.Mvc.Routing;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.AspNetCore.Routing.MapAction;
+
+namespace Microsoft.AspNetCore.Builder
+{
+    /// <summary>
+    /// Provides extension methods for <see cref="IEndpointRouteBuilder"/> to define HTTP API endpoints.
+    /// </summary>
+    public static class MapActionEndpointRouteBuilderExtensions
+    {
+        /// <summary>
+        /// Adds a <see cref="RouteEndpoint"/> to the <see cref="IEndpointRouteBuilder"/> that matches the pattern specified via attributes.
+        /// </summary>
+        /// <param name="endpoints">The <see cref="IEndpointRouteBuilder"/> to add the route to.</param>
+        /// <param name="action">The delegate executed when the endpoint is matched.</param>
+        /// <returns>A <see cref="IEndpointConventionBuilder"/> that can be used to further customize the endpoint.</returns>
+        public static IEndpointConventionBuilder MapAction(
+            this IEndpointRouteBuilder endpoints,
+            Delegate action)
```

Comment:
What's the plan to improve this?

Reply:
Source generators. That can generate the right overloads.

Reply:
I experimented with adding overloads that take `Func<T1, T2, .., TResult>`,  `Action<T1, T2, ...>`, etc... instead of `Delegate`, but you still need to cast the same way you do with the `Delegate` overload.

Overloads by themselves don't really give you anything today, and if we need compiler features anyway, why should it be necessary to source-gen the right overloads instead of continuing to use `Delegate`? I guess specific overloads could help optimize the startup code.

Reply:
The source generated model could also results in compile time code generation.